### ISSUE TITLE
hstack and zero-length arrays

### ISSUE THREAD

**Issue by RomanSteinberg** [numpy/numpy#7615]
I use the following code to widen masks (boolean 1D numpy arrays).
`mask = np.hstack([[False] * start, absent, [False]*rest])`
When start and rest are equal to zero, I've got an error, because mask becomes floating point 1D array. Generally speaking, when someone writes code like this:
`bool_a = np.hstack([bool_b, bool_c])`
he can get an error because semantics of stacking zero-length arrays with not zero-length assumes only floating point arrays in return. Actually, I haven't checked this code with arrays of type differ from bool and float. I think it is a bug.

**Comment by njsmith** [comment]
When you pass a list to a numpy function, you're relying on that function to automatically convert it to an array. The automatic array conversion code has to guess what type you meant, and for an empty list you're not giving it much to go on :-). What dtype should this have:
```
np.array([])
```
? Well, impossible to say, really...

In practice numpy defaults to float when it has no other clues, but you can override this by telling numpy what you actually want:
```
mask = np.hstack([np.asarray([False] * start, dtype=bool), absent, np.asarray([False]*rest, dtype=bool)])
```

**Comment by RomanSteinberg** [comment]
I agree that I should use implicit conversion more careful. But my point is that union of empty set with not empty set should not change the last one. I use set as a mathematical term here. It shouldn't guess what type of array is used it should do nothing with my data in non-empty array. 

I know tha default is float. If you think that my sample is correct, then explain why this semantics is ok.

**Comment by njsmith** [comment]
Type checking intentionally abstracts away from the actual contents of an array -- we check for type consistency before we do anything else.
```
# What should this return? Something empty, but what?
np.concatenate([np.array([], dtype=int), np.array([], dtype=float)])

a = np.array([1, 2, 3], dtype=int)
b = np.array([True, False, True], dtype=bool)
# This *could* be made to work when n is 0 or 3, but that would just be hiding the bug when n is 1 or 2
np.concatenate([a[:n], b[n:]])
```

**Comment by RomanSteinberg** [comment]
`a = np.array([1, 2, 3], dtype=int)` 
`b = np.array([True, False, True], dtype=bool)`
`np.concatenate([a[:n], b[n:]])`

In my opinion, numpy users who write such an ugly code should go to sell cakes instead of programming. This code will punch them right in the head when they start to use it. OR numpy can warn them. It depends on your decision! 
[Beautiful is better than ugly. Explicit is better than implicit.](https://www.python.org/dev/peps/pep-0020/)

> # What should this return? 
Error

We are wasting the time! If the numpy team thinks that my case is not an issue, ok I'll leave.

**Comment by eric-wieser** [comment]
> This code will punch them right in the head when they start to use it

What makes you say that? For all values of `n`, the result is of type `int`. That seems like correct behavior to me

> Beautiful is better than ugly. Explicit is better than implicit.

But _"Special cases aren't special enough to break the rules"_. Empty arrays aren't special, and should be treated the same way as any other size array.

Note that I say `array`s there. Perhaps we should warn if an empty `list` is passed in, and ask the user to promote to an array themselves

**Comment by RomanSteinberg** [comment]
> That seems like correct behavior to me

Concatenation of ints and bools in one array is ugly for me. I'm sure this code needs explicit conversion.

> Empty arrays aren't special, and should be treated the same way as any other size array.

Agreed. Empty arrays have the data type property and user responsible for choosing it oposite to lists. Warning for lists is looking good.

**Comment by eric-wieser** [comment]
> I'm sure this code needs explicit conversion.

Ah, I'm with you now. It sounds like your qualm is with `np.result_type`, which in general is perhaps a little too flexible (and is what `concatenate` falls back on)

Perhaps you should file an issue comparing the actual and desireable behaviour of `result_type`? Some examples you might not be pleased with

```
>>> np.result_type(np.int32, np.bool_)
dtype('int32')
>>> np.result_type(np.int32, np.str_)
dtype(<,U11')
```
Perhaps the best fix is to add a `casting` argument, so that a stricter `result_type(..., casting='safe')` can exist

**Comment by RomanSteinberg** [comment]
Well, sounds reasonable. But I'm not sure what do you suppose me to do. Should I open a new issue about  type casting in `np.result_type`?
